### PR TITLE
Optmise Disbursement layers on map

### PR DESCRIPTION
**Pull request by ajeety4**
## Product Description

Updates logic to create a single disbursement layer for lines connecting cases and mobile workers on map instead of one layer for each line. This significantly improves the map performance for higher assignments(>1k). 

## Technical Summary

[Ticket](https://dimagi.atlassian.net/browse/SC-4102)

## Feature Flag

Microplanning.

### Safety story

Behind a feature flag and changes only relates to disbursement.  Tested on local. 

### QA Plan

None

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

**Review by zandre-eng — APPROVED** [review]
Looks good! Just a minor comment.

File: corehq/apps/geospatial/static/geospatial/js/case_management.js [hunk 1]
```diff
@@ -214,6 +215,27 @@ hqDefine("geospatial/js/case_management", [
             });
         };
 
+        function addDisbursementLinesLayer(source) {
```
[the remark this answers is not in the document]
Comment (ajeety4) [reply corehq/apps/geospatial/static/geospatial/js/case_management.js]: Good point. I was in dilemma to move it earlier considering disbursement is specific to microplanning. However your point about layers makes good sense. 
Addressed in [here](https://github.com/dimagi/commcare-hq/pull/35516/commits/5f8dcd1531fd9ea521d26ff1a10ddacb5700c0e2)

**Comment by ajeety4** [comment]
> @ajeety4 Look good, but just one suggestion (feel free to evaluate and dispute it):
> 
> The `handleDisbursementResults` function iterates over all the users and cases, then the `connectUserWithCasesOnMap` iterates over all the users and cases again.
> 
> How about the following (note that I haven't tested it) - I removed the `userToCasesList` list to immediately construct the `disbursementLinesSource` (which would have been done in `connectUserWithCasesOnMap`), since we already have all the data we need to do so and we're skipping having to loop over all users/cases again:
> 
> ```
>         self.handleDisbursementResults = function (result) {
>             // Clean stale disbursement results
>             mapModel.removeDisbursementLayer();
> 
>             let groupId = 0;
>             let disbursementLinesSource = [];
>             Object.keys(result).forEach((userId) => {
>                 const user = mapModel.userMapItems().find((userModel) => {return userModel.itemId === userId;});
>                 mapModel.caseGroupsIndex[userId] = {groupId: groupId, item: user};
> 
>                 let userToCases = {'user': user, 'cases': []};
>                 mapModel.caseMapItems().forEach((caseModel) => {
>                     if (result[userId].includes(caseModel.itemId)) {
>                         const lineCoordinates = [
>                             [user.itemData.coordinates.lng, user.itemData.coordinates.lat],
>                             [caseModel.itemData.coordinates.lng, caseModel.itemData.coordinates.lat],
>                         ];
>                         disbursementLinesSource.features.push(
>                             {
>                                 type: 'Feature',
>                                 properties: {},
>                                 geometry: { type: 'LineString', coordinates: lineCoordinates },
>                             }
>                         );
>                         
>                         mapModel.caseGroupsIndex[caseModel.itemId] = {
>                             groupId: groupId,
>                             item: caseModel,
>                             assignedUserId: userId,
>                         };
>                     }
>                 });
>                 userToCasesList.push(userToCases);
>                 groupId += 1;
>             });
>             addDisbursementLinesLayer(disbursementLinesSource);
>             self.setBusy(false);
>         };
> ```


Hey Charl , this is great point.  The iteration is being done twice.
However the disbursement layer creation (and hence this source logic) is also called here from the review assignment .  So in the favor of DRY and cleaner code, I think the iteration is fine in this case.

While looking through it,  I did end up reusing the existing data structure and made some improvements. Changes are [here](https://github.com/dimagi/commcare-hq/pull/35516/commits/d0f4ca094e54110315107d75aee4062603394ad0)

File: corehq/apps/geospatial/static/geospatial/js/case_management.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (zandre-eng) [reply corehq/apps/geospatial/static/geospatial/js/case_management.js]: If you would like to keep the disbursement logic completely separate from the map model then you could make the function even more generic to only create a generic layer. Params can be passed to the function to define it (e.g. layer ID and paint options).

[the remark this answers is not in the document]
Comment (ajeety4) [reply corehq/apps/geospatial/static/geospatial/js/case_management.js]: That is a good idea. For now. I am good with the earlier suggestion and it is also more consistent with the other functions like `hasDisbursementLayer` and `removeDisbursementLayer`. 
I do think there are some things related to disbursement that can possibly be moved around however not sure if it is worth the effort.

**Review by zandre-eng — APPROVED** [review]

**Pull request merged**